File: src/app/bench/lib/multiplyMatrices.ts
const multiplyMatrices = (a: number[][], b: number[][]): number[][] | undefined => {
    if (a[0].length !== b.length) return; // check matrix multiplication validity
    const result: number[][] = Array.from({ length: a.length }, () => Array.from({ length: b[0].length }, () => 0));
    
    for (let i = 0; i < result.length; i++) {
      for (let j = 0; j < result[0].length; j++) {
        for (let k = 0; k < a[0].length; k++) {
          result[i][j] += a[i][k] * b[k][j];
        }
      }
    }
  
    return result;
  };
  
  export {
    multiplyMatrices
  }

File: wasm/multiplyMatrices.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void generate_matrix(int n, int matrix[n][n]) {
    for(int i = 0; i < n; i++)
        for(int j = 0; j < n; j++)
            matrix[i][j] = rand() % 10;
}

void multiply_matrices(int n, int a[n][n], int b[n][n], int result[n][n]) {
    for(int i = 0; i < n; i++)
        for(int j = 0; j < n; j++) {
            result[i][j] = 0;
            for(int k = 0; k < n; k++)
                result[i][j] += a[i][k] * b[k][j];
        }
}

double benchmark(int n) {
    int a[n][n], b[n][n], result[n][n];

    srand(time(0));  // Seed for random number generator
    generate_matrix(n, a);
    generate_matrix(n, b);

    clock_t start = clock();
    multiply_matrices(n, a, b, result);
    clock_t end = clock();
    double time_taken = ((double)end - start) / CLOCKS_PER_SEC; // in seconds

    return time_taken;
}
